### ISSUE TITLE
`matmul` 

### ISSUE THREAD

**Issue by scarlehoff** [mhostetter/galois#500]

**Comment by mhostetter** [comment]
I don't have any good WIP code to share.

The thing I quickly discovered was CuPy arrays are not subclasses of `np.ndarray`. So the overridden hooks in `FieldArray` wouldn't trigger on a CuPy array. I'll need some other solution for a `FieldArray` on GPU. But these are issues with the API and interface for users.

Regarding the algorithm porting, it should be doable. But the API might not be as pretty.

What's your use-case field? I can point you to some small changes that might enable testing on a GPU.

**Comment by scarlehoff** [comment]
An use-case example would be:
```python
import galois
import numpy as np

p = 2**31 - 19

FF = galois.GF(p)

ar1 = np.random.randint(1, p, size=40).reshape(4, 5, 2)
ar2 = np.random.randint(1, p, size=90).reshape(5, 9, 2)

f1 = FF(ar1)
f2 = FF(ar2)

final_result = np.tensordot(f1, f2, axes=[(2, 1), (2, 0)])
```

Our preferred use-case would be something like this. Of course this could also be achieved with

```python
t1 = f1.reshape(4, 10)
t2 = np.transpose(f2, (0,2,1)).reshape(10, 9)
tt = t1 @ t2
```
so `matmul` is enough (`tensordor` would just be "more convenient", `einsum` even better :P)

We have a batch dimension so one of the axis would be some large number of events (hence the benefit from the gpu)

**Comment by mhostetter** [comment]
With the exception of the `tensordot()` call, the above code currently runs. However, it's on CPU. I'm guessing in your desired use case you'd use CuPy?

Also, a tip: If you have large arrays, it's better to use `f1 = ar1.view(FF)` rather than `f1 = FF(ar1)`. `view()` is copy-less while `FF()` uses a copy. See https://mhostetter.github.io/galois/latest/basic-usage/array-creation/#numpy-array

**Comment by scarlehoff** [comment]
> I'm guessing in your desired use case you'd use CuPy?

Yes, `CuPy` would be nice. But we are very flexible in the choice of library (the more numpy-like the interface is the better, of course, but it is just a minor inconvenience).
